x/warp/config.go

```
@@ -0,0 +1,213 @@
+	}
+	return nil
+}
+
+// Equal returns true if [s] is a [*Config] and it has been configured identical to [c].
+func (c *Config) Equal(s precompileconfig.Config) bool {
+	// typecast before comparison
+	other, ok := (s).(*Config)
+	if !ok {
+		return false
+	}
+	equals := c.Upgrade.Equal(&other.Upgrade)
+	return equals && c.QuorumNumerator == other.QuorumNumerator
+}
+
+func (c *Config) Accept(acceptCtx *precompileconfig.AcceptContext, txHash common.Hash, logIndex int, topics []common.Hash, logData []byte) error {
+	unsignedMessage, err := warp.ParseUnsignedMessage(logData)
+	if err != nil {
+		return fmt.Errorf("failed to parse warp log data into unsigned message (TxHash: %s, LogIndex: %d): %w", txHash, logIndex, err)
+	}
+	log.Info("Accepted warp unsigned message", "txHash", txHash, "logIndex", logIndex, "logData", common.Bytes2Hex(logData))
+	if err := acceptCtx.Warp.AddMessage(unsignedMessage); err != nil {
+		return fmt.Errorf("failed to add warp message during accept (TxHash: %s, LogIndex: %d): %w", txHash, logIndex, err)
+	}
+	return nil
+}
+
+// verifyWarpMessage checks that [warpMsg] can be parsed as an addressed payload and verifies the Warp Message Signature
+// within [predicateContext].
+func (c *Config) verifyWarpMessage(predicateContext *precompileconfig.ProposerPredicateContext, warpMsg *warp.Message) error {
+	// Use default quourum numerator unless config specifies a non-default option
+	quorumNumerator := params.WarpDefaultQuorumNumerator
+	if c.QuorumNumerator != 0 {
+		quorumNumerator = c.QuorumNumerator
+	}
+
+	// Verify the warp payload can be decoded to the expected type
+	_, err := warpPayload.ParseAddressedPayload(warpMsg.UnsignedMessage.Payload)
+	if err != nil {
+		return fmt.Errorf("%w: %s", errInvalidAddressedPayload, err)
+	}
+
+	log.Info("verifyingWarpMessage", "warpMsg", warpMsg, "quorumNum", quorumNumerator, "quorumDenom", params.WarpQuorumDenominator)
+	if err := warpMsg.Signature.Verify(
```

Should we fail verification when `warpMsg.Signature.NumSigners() == 0` before calling `.Verify(...)`?

---

This is in predicate gas, which should only fail imo if it is unable to calculate the amount of gas, not if the message should be invalid. The call to `Verify(...)` should perform operations in the correct order, such that if there is a sign the message will fail verification, we will minimize wasted work, but I don't think we should try to fail early within `PredicateGas(...)`

---

The codec will fail to unmarshal if it's going to return a nil value.